File: leetcode/parenthesis/1249. Minimum Remove to Make Valid Parentheses/js/main.js
// Time O(N)
// Space O(N)
/**
 * @param {string} s
 * @return {string}
 */
const minRemoveToMakeValid = s => {
  let st = [];
  let removed = new Set();

  for (let i = 0; i < s.length; i++) {
    if (s[i] == '(') {
      st.push(i);
    } else if (s[i] == ')') {
      if (st.length == 0) {
        removed.add(i);
      } else {
        st.pop();
      }
    }
  }

  while (st.length) {
    removed.add(st.pop());
  }

  let res = '';

  for (let i = 0; i < s.length; i++) {
    if (!removed.has(i)) {
      res += s[i];
    }
  }

  return res;
};

File: leetcode/parenthesis/1249. Minimum Remove to Make Valid Parentheses/py/main.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        st = []
        removed = set()

        for i, ch in enumerate(s):
          if ch == '(':
            st.append(i)
          elif ch == ')':
            if len(st) == 0:
              removed.add(i)
            else:
              st.pop()
        
        while len(st):
          removed.add(st.pop())

        res = ''

        for i, ch in enumerate(s):
          if i not in removed:
            res += ch

        return res